manager/4337/README.md

```diff
@@ -0,0 +1,114 @@
+
+Storage associated with an account is defined as follows:
+
+> An address `A` is associated with:
+>
+> - Slots of contract `A` address itself.
+> - Slot `A` on any other address.
+> - Slots of type `keccak256(A || X) + n` on any other address. (to cover `mapping(address => value)`, which is usually used for balance in ERC-20 tokens). `n` is an offset value up to 128, to allow accessing fields in the format `mapping(address => struct)`
+
+As such, in order for the Safe{Core} Protocol to be ERC-4337 compatible without requiring a staked _paymaster_, the Manager MUST NOT make use of storage slots not associated with the account whose `UserOperation`s it is validating. Notably, this implies that `UserOperation` validation MUST be implemented without interacting with the Registry. If ERC-4337 `validateUserOp` support is implemented as a module, the Manager MUST NOT verify it is approved in the Registry. Note that this restriction only applies to `UserOperation` validation flow and notably does not apply to the `UserOperation` execution flow. The specification currently suggests that ERC-4337 `validateUserOp` SHOULD be directly supported in the Manager for these compatability reasons.
+
+Additionally, developers aiming to develop plugins that are ERC-4337 compatible should be aware of storage access restrictions, opcode usage restrictions during the simulation step of ERC-4337 specification.
+
+More details on this is available [here](https://github.com/safe-global/safe-core-protocol/issues/60#issuecomment-1761296305).
+
+### Staking
+
+The account SHOULD implement a staked _factory_ in order to enable `UserOperations` with `initCode`. ERC-4337 requires a staked _factory_ for `UserOperation` validation that reads associated account storage when `initCode` is specified. Since this specification permits Manager implementations that read account associated storage during `UserOperation` validation, the account _factory_ needs to be staked in order for bundlers to accept `UserOperations` with `initCode`.
+
+## Security Considerations
+
+One of the main value propositions of the Safe{Core} Protocol is account security by verifying modules in the Registry. However, because of storage access restrictions imposed by the ERC-4337 spec (see Compatibility section above), this is not possible when validating user operations. However, this restriction does not apply to `UserOperation` execution.
+
+In order to help protect accounts from any eventual vulnerabilities that may be discovered in the ERC-4337 _entrypoint_, The manager MUST validate `UserOperation` execution with the registry for the particular _entrypoint_ being used. Practically speaking this implies that the ERC-4337 `UserOperation` execution:
+1. SHOULD be implemented as a fallback handler module with an immutable associated _entrypoint_ contract
+2. MUST verify that the _entrypoint_ has been explicitely enabled by the account
+3. MUST verify that the _entrypoint_ has not been flagged
```

Comment:
do the storage read rules permit this? Or how will it work?

Reply:
This is in `executeUserOp` (i.e. post-`UserOperation` validation in `validateUserOp`), so AFAIU there are no storage restrictions here.

What could happen if the `Registry` has flagged the entrypoint would be that:
1. The user's `validateUserOp` still succeeds, and the account pays the prefund
2. The user operation would revert instead of executing

So, while this is not ideal (the gas is spent), the account is still slightly more protected in case of:
1. A vulnerability in the `EntryPoint` allowing for arbitrary `UserOperation` execution
2. A vulnerability in the 4337 plugin, that would allow for arbitrary `UserOperation` execution.

It isn't fully protected, however, because `validateUserOp` still transfers `missingAccountFunds` to `msg.sender`, but it does prevent the account from executing arbitrary transactions in case of a bug somewhere.

Reply:
Ah, understood, I missed the word "execution" in the specs